Zero-copy creation of `Address` objects

jakirkham opened the pull request:
When setting or getting the underlying string data used for an `Address` object, avoid extra copying. This is done by using `string_view` where possible in C++ & Cython and by using `bytes`-like objects in Python. This cuts down on the allocation and copying of memory for temporary strings and deferring them to the relevant point where an allocation and copy must happen for the owning object.

Users remain free to copy and persist the `Address`'s as a `std::string` or `bytes` object at any point. This merely tries to defer that decision to the user instead of making it for them

jakirkham committed:
Use `string_view` in address methods

When setting or getting the underlying string data used for an `Address`
object, avoid extra copying by using `string_view`. During
initialization a copy is performed into the `Address` object's local
stroage, which makes sense. Though the input to the function need not
copy to `string`. It can just be viewed while the data is copied into
the object's local memory. Further when converting the `Address` to a
string, a view can be used to avoid a copy during the return. If the
caller would like to persist the string longer than the lifetime of the
`Address` object, they can copy it at that point. Though we can save
them a copy before then. Especially if they will use the string while
the `Address` object is present.

jakirkham committed:
Reuse `length` assignment in allocation

jakirkham committed:
Bump license header year range

pentschev commented:
Thanks for submitting this John! Unsurprisingly the tests failed:

```
ImportError: /opt/conda/conda-bld/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib/python3.12/site-packages/ucxx/_lib/libucxx.cpython-312-x86_64-linux-gnu.so: undefined symbol: _ZN4ucxx23createAddressFromStringENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
```

That's because the Cython API has not been updated as well. Should we make all the necessary changes here, including bumping Cython to 3.1 and leave the PR ready for the time Cython 3.1 is out?

jakirkham commented:
Ah sometimes the Cython API will let these things slide as it is usually just checking that something will work with the underlying C++ API (as opposed to enforcing it)

We could certainly do that

For now would it make sense to use `const std::string&`? That would save us some copies between these API calls

wence- commented on cpp/src/address.cpp:
```diff
@@ -38,21 +38,21 @@ std::shared_ptr<Address> createAddressFromWorker(std::shared_ptr<Worker> worker)
-std::shared_ptr<Address> createAddressFromString(std::string addressString)
+std::shared_ptr<Address> createAddressFromString(std::string_view addressString)
 {
-  ucp_address_t* address = reinterpret_cast<ucp_address_t*>(new char[addressString.length()]);
   size_t length          = addressString.length();
-  memcpy(address, addressString.c_str(), length);
+  ucp_address_t* address = reinterpret_cast<ucp_address_t*>(new char[length]);
+  memcpy(address, addressString.data(), length);
   return std::shared_ptr<Address>(new Address(nullptr, address, length));
 }
```
This is fine.

jakirkham replied:
For additional context, this is used in one place in Cython where we already have something that could be `std::string_view` compliant (a Python `bytes` object). Currently in the construction we...

1. Copy the `bytes` object into a `std::string`
2. Copy the `std::string` to the `createAddressFromString` function call (no `const std::string&` currently)
3. Allocate and copy the `std::string` to a `char[]`

https://github.com/rapidsai/ucxx/blob/4cd6354fc23e62f19b9715c04fd69dd1a1298997/python/ucxx/ucxx/_lib/libucxx.pyx#L407-L416

4. Note there is a `getString` call at the end doing one more copy

wence- commented on cpp/src/address.cpp:
```diff
@@ -38,21 +38,21 @@ std::shared_ptr<Address> createAddressFromWorker(std::shared_ptr<Worker> worker)
 size_t Address::getLength() const { return _length; }
 
-std::string Address::getString() const
+std::string_view Address::getString() const
 {
-  return std::string{reinterpret_cast<char*>(_handle), _length};
+  return std::string_view{reinterpret_cast<const char*>(_handle), _length};
 }
```
NAK, this is horrible because now I have to remember to tie the lifetime of the returned object to the `Address`. _And_ the docstring is now wrong.

I cannot believe this is worth it.

pentschev replied:
This is a good point, we can make it the user's responsibility to ensure the lifetime, [this is already the case for the UCX handles](https://github.com/rapidsai/ucxx/blob/4cd6354fc23e62f19b9715c04fd69dd1a1298997/cpp/include/ucxx/address.h#L83-L86). I don't have a strong objection either way, I'm a little more inclined towards Lawrence's point simply because that would avoid the potential to dereference an invalid pointer, whereas if attempting to use the address (that has become invalid because the worker is not available anymore) would simply result in an error from UCX/exception from UCXX.

@wence- @jakirkham do you have strong opinions?

jakirkham replied:
AFAICT there are only two uses of this method. Both involve the creation of a Cython `UCXAddress` object

1. From `bytes` as described above (currently 4 allocations & copies): https://github.com/rapidsai/ucxx/pull/392#discussion_r2034379410
2. From the Cython `UCXWorker` described below

In the Cython `UCXWorker` case, we take and store a copy of the `std::shared_ptr<Address>`. The `Address` in turn owns and keeps alive the `ucp_address_t*`/`char*` representation. Additionally we...

1. Allocate and copy `ucp_address_t*` to `std::string`
2. Copy this `std::string` into the Cython `UCXAddress`

https://github.com/rapidsai/ucxx/blob/4cd6354fc23e62f19b9715c04fd69dd1a1298997/python/ucxx/ucxx/_lib/libucxx.pyx#L395-L405

In both cases we are doing a fair bit of additional object creation and copying. The goal of this PR was to cutdown on that

<hr>

Though would acknowledge that

1. Many ways to achieve the same goal
2. Some amount of creation and copying may be needed

That said, think it would be good if we could reduce the object creation and copying. Ideally deferring copying until the final object is constructed

<hr>

Side note: `std::string` can be trivially constructed from a `std::string_view`

jakirkham replied:
That said, if we would rather call this behavior `getStringView` that also seems reasonable

jakirkham replied:
Should add the docstring has been updated. So think this is addressed

Happy to discuss more as needed

pentschev commented:
> For now would it make sense to use `const std::string&`? That would save us some copies between these API calls

Yes, using `const std::string&` immediately sounds appropriate to me

jakirkham committed:
Merge branch 'main' into mk_add_sv

jakirkham committed:
Bump license headers for 2026

jakirkham committed:
Def `createAddressFromString` with `string_view`

jakirkham committed:
Def Cython createAddressFromString w/string_view

jakirkham committed:
Use `std::string_view` in Cython too

Instead of constructing a `std::string` from a Python `bytes` object,
construct a `std::string_view`. Then use that when calling
`createAddressFromString`. This avoids allocating and copying the same
string a few times. Instead it simply forwards the string to use to the
code site where the `Address` is created. As `Address` needs to retain
its own string and memory, it copies, but nothing before it needs to.

jakirkham committed:
Update `getString` docs to note a view is returned

jakirkham committed:
Accept bytes-like objects in create_from_buffer

Previously `create_from_buffer` only accepted `bytes`. As a result
anything passing an object to `create_from_buffer` either needed to
ensure it was recieving a `bytes` object or it needed to create a
`bytes` object from anything it had before passing it to
`create_from_buffer`. By relaxing `create_from_buffer` to allow any
`bytes`-like object, callers can continue to pass `bytes` objects when
they have them or they can pass anything that looks similar (like
`bytearray` or NumPy 1-D `uint8` arrays). This can save additional
allocations and copies upstream as well.

jakirkham committed:
Test `create_from_buffer` with `Address` objects

Ensure `Address` objects can be passed unchanged to
`create_from_buffer`. Under-the-hood it copies the address as expected.
Test that this holds.

jakirkham committed:
Test `bytearray`s in `create_from_buffer`

Ensure other `bytes`-like objects can be handled by `create_from_buffer`
without needing to convert them to `bytes` first.

jakirkham committed:
Pass input through `get_ucx_address_from_buffer`

Skip the `bytes` conversion within `get_ucx_address_from_buffer` as
`create_from_buffer` can accept any `bytes`-like object now. This save
another allocation and copy as it can pass the address through.

jakirkham committed:
Bump more years to 2026

jakirkham committed:
Provide zero-copy serialization of `Address`

If pickle protocol 5 is available, pass `Address` via `PickleBuffer` to
avoid unnecessary copies during serialization. Otherwise fallback to
converting `Address` to `bytes`.

jakirkham committed:
Note `getString` is still returning a string

jakirkham committed:
Drop extra `protocol` argument in `loads`

jakirkham committed:
Merge branch 'main' into mk_add_sv

wence- commented on python/ucxx/ucxx/_lib/libucxx.pyx:
```diff
@@ -581,8 +583,11 @@ cdef class UCXAddress():
+            return (UCXAddress.create_from_buffer, (PickleBuffer(self),))
+        else:
+            return (UCXAddress.create_from_buffer, (bytes(self),))
 
```
Why use a `PickleBuffer`? AFAICT it's basically just a `bytes`?

jakirkham committed:
Merge branch 'main' into mk_add_sv

jakirkham committed:
Merge branch 'main' into mk_add_sv

jakirkham committed:
Merge branch 'main' into mk_add_sv

jakirkham committed:
Merge branch 'main' into mk_add_sv

pentschev committed:
Merge branch 'main' into mk_add_sv

pentschev reviewed (approved):
I think this looks good now, thanks John!

jakirkham commented:
Thanks Peter and Lawrence! 🙏

Am planning to add a few tests for this as well. Will do that next week

jakirkham committed:
Merge branch 'main' into mk_add_sv

jakirkham committed:
Test serialized content before deserializing

jakirkham committed:
Test creation of `UCXAddress` from `memoryview`s

jakirkham committed:
Retitle pickle protocol 5 test of `UCXAddress`

jakirkham committed:
Test `PickleBuffer` creation of `UCXAddress`

jakirkham commented:
Have included a few more tests

Changes are done from my end

Happy to discuss anything else as needed

pentschev reviewed (approved):
Thanks John!

jakirkham committed:
Update Cython's C++ bindings to use string_view

jakirkham committed:
Explicitly create `std::string` for address

jakirkham committed:
Tweak docstring to note `std::string` creation

jakirkham committed:
Use explicit `std::string` in `create_from_worker`

jakirkham committed:
Use `std::string_view` for `UCXAddress._string`

jakirkham commented on python/ucxx/ucxx/_lib/libucxx.pxd:
```diff
@@ -42,7 +42,7 @@ cdef class UCXAddress:
         shared_ptr[Address] _address
         size_t _length
         ucp_address_t *_handle
-        string _string
+        string_view _string
```
Realized this could also just be a `string_view` since it is just viewing the memory owned by `_address`. The same memory that the pointer `_handle` refers to

This attribute is only internally used. So it just avoids some copies when working with internal data in class methods

jakirkham committed:
Update `libucxx.pxd` for 2026

pentschev reviewed (approved):
Still looking good, thanks John. We can leave this open until EOD in case there are further comments and hit merge if not.

pentschev commented:
/merge

pentschev commented:
I just realized this is in fact breaking, the change of `getString()`'s return type _is breaking_.

jakirkham commented:
Thanks Peter! 🙏

Ah thank you for catching that. Had done a pretty extensive search of the RAPIDS code base and only found this function used in UCXX and in that Cython code. So not expecting other impact. That said, agree that it is more accurate to say this is a breaking change

If we would rather have a deprecation cycle for good measure, am open to doing that. Just let me know

Edit: Also asking OPS if we can still have this marked as breaking in the changelog

jakirkham commented:
> If we would rather have a deprecation cycle for good measure, am open to doing that. Just let me know

Went ahead and put something together in case we want it in PR: https://github.com/rapidsai/ucxx/pull/622

Ok either way

pentschev commented:
> Thanks Peter! 🙏
> 
> Ah thank you for catching that. Had done a pretty extensive search of the RAPIDS code base and only found this function used in UCXX and in that Cython code. So not expecting other impact. That said, agree that it is more accurate to say this is a breaking change
> 
> If we would rather have a deprecation cycle for good measure, am open to doing that. Just let me know
> 
> Edit: Also asking OPS if we can still have this marked as breaking in the changelog

I only found that because [we use it in RapidsMPF as well](https://github.com/rapidsai/rapidsmpf/blob/6103323e9dc954095ccd63c0405bca5f0ad6f717/cpp/src/bootstrap/ucxx.cpp#L47-L50) and it was using `auto` but passing to a function that expected a `std::string`, which in the linked PR was adjusted to accept `std::string_view` now. Yesterday I pondered whether we needed some deprecation cycle and though we maybe don't, the amount of work will be relatively the same for users and we don't (yet) guarantee API/ABI stability.

jakirkham commented:
Ah sorry for the disruption there and not catching those previously. Thank you for following up on those issue and letting me know! 🙏

Will take a closer look at RapidsMPF going forward
